### PR TITLE
openthread: add support for radios compatible with SubMAC

### PR DESCRIPTION
**Pull request by jia200x**
### Contribution description
This PR adds OpenThread support for the radios compatible with the SubMAC
This work requires the `netdev_ieee802154_submac_t` interface of IEEE 802.15.4 SubMAC (#14950 ) to work.

This acts as a showcase for the IEEE 802.15.4 Radio HAL and SubMAC, since all APIs are hardware independent now. Thus, it's possible to transparently integrate OpenThread with any radio that implements the IEEE 802.15.4 Radio HAL.

In the future, we could drop the `netdev_ieee802154_submac_t` interface and implement the platform abstraction directly on top of the Radio HAL or SubMAC.

### Testing procedure
Run `make -C examples/openthread` with any CC2538 based board (e.g remote-revb).

### Issues/PRs references
Now this depends on #15237 
#10045 
#13376

### CODE REVIEW

**Comment by benpicco** [comment]
Needs a rebase now :smiley:

**Comment by jia200x** [comment]
> Also the title is confusing. This should not have much to do with CC2538.

Yes, you are right. It was just because it was the only radio available at that time

**Comment by jia200x** [comment]
Done. I also updated the title to reflect the changes.

File: pkg/openthread/Makefile.dep [hunk 1]
```diff
@@ -4,4 +4,10 @@ USEMODULE += openthread_contrib_netdev
 USEMODULE += l2util
 USEMODULE += xtimer
 USEMODULE += event
+USEMODULE += netdev_default
+
+ifneq (,$(filter cc2538_rf,$(USEMODULE)))
```
Comment (benpicco) [inline pkg/openthread/Makefile.dep:9]: You can add `nrf802154` to the list

**Review by benpicco — COMMENTED** [review]
looks like a simple addition now - please rebase

File: pkg/openthread/Makefile.dep (hunk 1 above)
Comment (jia200x) [reply pkg/openthread/Makefile.dep:9]: true! I will

**Comment by jia200x** [comment]
done!

**Comment by aabadie** [comment]
openthread will only work on boards that using uart as stdio. Some nrf52840 based boards (adafruit-clue at least) uses USB CDC ACM as stdio. Is it possible to use `stdio_write` instead of `uart_write` in the openthread port ? That could solve that issue.

**Comment by aabadie** [comment]
For example, it's not possible to build `examples/openthread` on `adafruit-clue` with the current state of this PR:

<details>

```
BUILD_IN_DOCKER=1 make BOARD=adafruit-clue -C examples/openthread --no-print-directory 
Launching build container using image "riot/riotbuild:latest".
docker run --rm --tty --user $(id -u) -v '/usr/share/zoneinfo/Europe/Paris:/etc/localtime:ro' -v '/work/riot/RIOT:/data/riotbuild/riotbase:delegated' -e 'RIOTBASE=/data/riotbuild/riotbase' -e 'CCACHE_BASEDIR=/data/riotbuild/riotbase' -e 'BUILD_DIR=/data/riotbuild/riotbase/build' -e 'RIOTPROJECT=/data/riotbuild/riotbase' -e 'RIOTCPU=/data/riotbuild/riotbase/cpu' -e 'RIOTBOARD=/data/riotbuild/riotbase/boards' -e 'RIOTMAKE=/data/riotbuild/riotbase/makefiles'      -e 'BOARD=adafruit-clue'  -w '/data/riotbuild/riotbase/examples/openthread/' 'riot/riotbuild:latest' make 'BOARD=adafruit-clue'    
Building application "openthread" for "adafruit-clue" with MCU "nrf52".

"make" -C /data/riotbuild/riotbase/pkg/openthread
"make" -C /data/riotbuild/riotbase/boards/adafruit-clue
"make" -C /data/riotbuild/riotbase/core
"make" -C /data/riotbuild/riotbase/cpu/nrf52
"make" -C /data/riotbuild/riotbase/cpu/cortexm_common
"make" -C /data/riotbuild/riotbase/cpu/cortexm_common/periph
"make" -C /data/riotbuild/riotbase/cpu/nrf52/periph
"make" -C /data/riotbuild/riotbase/cpu/nrf52/radio/nrf802154
"make" -C /data/riotbuild/riotbase/cpu/nrf52/vectors
"make" -C /data/riotbuild/riotbase/cpu/nrf5x_common
"make" -C /data/riotbuild/riotbase/cpu/nrf5x_common/periph
"make" -C /data/riotbuild/riotbase/drivers
"make" -C /data/riotbuild/riotbase/drivers/netdev
"make" -C /data/riotbuild/riotbase/drivers/netdev_ieee802154_submac
"make" -C /data/riotbuild/riotbase/drivers/periph_common
"make" -C /data/riotbuild/riotbase/pkg/openthread/contrib
"make" -C /data/riotbuild/riotbase/pkg/openthread/contrib/netdev
"make" -C /data/riotbuild/riotbase/sys
"make" -C /data/riotbuild/riotbase/sys/auto_init
"make" -C /data/riotbuild/riotbase/sys/auto_init/usb
"make" -C /data/riotbuild/riotbase/sys/cpp11-compat
"make" -C /data/riotbuild/riotbase/sys/div
"make" -C /data/riotbuild/riotbase/sys/event
"make" -C /data/riotbuild/riotbase/sys/isrpipe
"make" -C /data/riotbuild/riotbase/sys/luid
"make" -C /data/riotbuild/riotbase/sys/net/link_layer/eui_provider
"make" -C /data/riotbuild/riotbase/sys/net/link_layer/ieee802154
"make" -C /data/riotbuild/riotbase/sys/net/link_layer/l2util
"make" -C /data/riotbuild/riotbase/sys/newlib_syscalls_default
"make" -C /data/riotbuild/riotbase/sys/ps
"make" -C /data/riotbuild/riotbase/sys/random
"make" -C /data/riotbuild/riotbase/sys/random/tinymt32
"make" -C /data/riotbuild/riotbase/sys/timex
"make" -C /data/riotbuild/riotbase/sys/tsrb
"make" -C /data/riotbuild/riotbase/sys/usb/usbus
"make" -C /data/riotbuild/riotbase/sys/usb/usbus/cdc/acm
"make" -C /data/riotbuild/riotbase/sys/usb_board_reset
"make" -C /data/riotbuild/riotbase/sys/xtimer
/opt/gcc-arm-none-eabi-9-2019-q4-major/bin/../lib/gcc/arm-none-eabi/9.2.1/../../../../arm-none-eabi/bin/ld: /data/riotbuild/riotbase/examples/openthread/bin/adafruit-clue/openthread_contrib/platform_uart.o: in function `otPlatUartEnable':
/data/riotbuild/riotbase/pkg/openthread/contrib/platform_uart.c:83: undefined reference to `uart_init'
/opt/gcc-arm-none-eabi-9-2019-q4-major/bin/../lib/gcc/arm-none-eabi/9.2.1/../../../../arm-none-eabi/bin/ld: /data/riotbuild/riotbase/examples/openthread/bin/adafruit-clue/openthread_contrib/platform_uart.o: in function `otPlatUartSend':
/data/riotbuild/riotbase/pkg/openthread/contrib/platform_uart.c:97: undefined reference to `uart_write'
collect2: error: ld returned 1 exit status
/data/riotbuild/riotbase/Makefile.include:594: recipe for target '/data/riotbuild/riotbase/examples/openthread/bin/adafruit-clue/openthread.elf' failed
make: *** [/data/riotbuild/riotbase/examples/openthread/bin/adafruit-clue/openthread.elf] Error 1
```

</details>

**Comment by jia200x** [comment]
> openthread will only work on boards that use uart as stdio. Some nrf52840 based boards (adafruit-clue at least) uses USB CDC ACM as stdio. Is it possible to use stdio_write instead of uart_write in the openthread port ? That could solve that issue.

Yes! That shoul dbe possible. I will check

**Comment by jia200x** [comment]
> Yes! That should be possible. I will check

Would it be OK if I remove those boards and then adapt OpenThread for such changes? Since it involves stuff with the contrib files

**Comment by benpicco** [comment]
> Would it be OK if I remove those boards and then adapt OpenThread for such changes? Since it involves stuff with the contrib files

Just make a separate PR for it, we could probably merge that quick.

File: examples/openthread/Makefile [hunk 2]
```diff
@@ -4,7 +4,31 @@ APPLICATION = openthread
...
+					cc2538dk \
+					remote-reva \
+					remote-revb \
+					omote \
+					openmote-b \
+					openmote-cc2538 \
+					adafruit-clue \
+					adafruit-itsybitsy-nrf52 \
+					arduino-nano-33-ble \
+					feather-nrf52840 \
```
Comment (aabadie) [inline examples/openthread/Makefile:26]: Those boards should not be part of this PR because they don't work with openthread (no stdio on uart).
```suggestion
```

File: examples/openthread/Makefile [hunk 3]
```diff
@@ -4,7 +4,31 @@ APPLICATION = openthread
...
+					nrf52840dk \
+					nrf52840dongle \
```
Comment (aabadie) [inline examples/openthread/Makefile:28]: Same for this one:
```suggestion
```

**Comment by aabadie** [comment]
> Just make a separate PR for it, we could probably merge that quick.

No need for a separate PR, the boards are whitelisted by this one so they should just be removed from the white list here.

**Comment by jia200x** [comment]
done!

**Comment by aabadie** [comment]
But it would be nice to have a follow-up PR that improves the situation with openthread + stdio on USB CDC ACM and enables them.

**Comment by benpicco** [comment]
Feel free to squash.

**Comment by jia200x** [comment]
done!

File: pkg/openthread/Makefile.dep [hunk 4]
```diff
@@ -4,4 +4,10 @@ USEMODULE += openthread_contrib_netdev
 USEMODULE += l2util
 USEMODULE += xtimer
 USEMODULE += event
+USEMODULE += netdev_default
```
Comment (benpicco) [inline pkg/openthread/Makefile.dep:7]: That shouldn't be up to the pkg to choose.
Just remove that line, it's usually set by application (and squash again)

**Comment by jia200x** [comment]
done!

**Review by benpicco — APPROVED** [review]

**Comment by benpicco** [comment]
Murdock is not happy

**Comment by jia200x** [comment]
I added an extra commit to add `netdev_default` in `openthread/examples` (see https://github.com/RIOT-OS/RIOT/pull/14979/commits/edc4d1df340968e077c3997be7f71d3831a745d2). Someone has to pull the radio anyway.

**Comment by aabadie** [comment]
> I added an extra commit to add netdev_default in openthread/examples

Why in the application and not in the package ? Since OpenThread is network stack that lives directly on top of interfaces, that would make more sense and simplify the application Makefile.

**Comment by benpicco** [comment]
`netdev_default` just says you want the default netdev for a board.
If you have a different radio connected to your board or want to try a different driver you would remove that and replace it with your other radio driver.

That's how `examples/gnrc_networking` does it too (and `grep netdev_default examples/*/Makefile` shows it's not the only one)

File: examples/openthread/Makefile [hunk 5]
```diff
@@ -4,7 +4,26 @@ APPLICATION = openthread
...
+					frdm-kw41z \
+					openlabs-kw41z-mini-256kib \
+					openlabs-kw41z-mini \
+					phynode-kw41z \
+					usb-kw41z \
+					cc2538dk \
+					remote-reva \
+					remote-revb \
+					omote \
+					openmote-b \
```
Comment (benpicco) [inline examples/openthread/Makefile:21]: OpenMote-B defaults to `at86rf215` which has not yet been converted to the radio HAL.

File: examples/openthread/Makefile [hunk 6]
```diff
@@ -4,7 +4,26 @@ APPLICATION = openthread
 BOARD ?= samr21-xpro
 
 # These are the boards that OpenThread stack has been tested on
-BOARD_WHITELIST := samr21-xpro iotlab-m3 fox iotlab-a8-m3 frdm-kw41z openlabs-kw41z-mini-256kib openlabs-kw41z-mini phynode-kw41z usb-kw41z
+BOARD_WHITELIST := \
+					samr21-xpro \
```
Comment (benpicco) [inline examples/openthread/Makefile:8]: Why introduce so much Whitespace?

Comment (jia200x) [reply examples/openthread/Makefile:8]: I think this was because of the tabs. I will remove them

File: examples/openthread/Makefile (hunk 5 above)
Comment (jia200x) [reply examples/openthread/Makefile:21]: done!

File: examples/openthread/Makefile (hunk 6 above)
Comment (jia200x) [reply examples/openthread/Makefile]: done!
